File: scenario.py
from finish_conditions import IConditionsChecker
from runner import IRunner


class Scene:
    def __init__(self, cond: IConditionsChecker, runner: IRunner) -> None:
        self.cond = cond
        self.runner = runner

    def run(self) -> bool:
        self.runner.run()
        return self.cond.check()


class Scenario:

    def __init__(self) -> None:
        self.scenes: list[Scene] = []
        self.scene_index = 0

    def run(self) -> bool:
        if self.scene_index >= len(self.scenes):
            return True
        result = self.scenes[self.scene_index].run()
        if result:
            self._change_scene()
        return False

    def _change_scene(self):
        self.scene_index += 1

    def append(self, scene: Scene):
        self.scenes.append(scene)


class ScenarioTracer:
    def run(self) -> bool:
        pass

File: src/scenario_control/Scenario.cpp
#include "scenario_control/Scenario.hpp"

void scenario_control::Scenario::append(scenario_control::Scene &scene) {
  this->scenes.push_back(scene);
}

bool scenario_control::Scenario::run() {
  if (this->scenes.size() == 0 ||
      this->current_scene_index >= this->scenes.size()) {
    return true;
  }

  bool result = this->scenes[this->current_scene_index].run();
  if (result) {
    this->change_scene();
  }

  return false;
}

void scenario_control::Scenario::change_scene() { this->current_scene_index++; }